File: ruby/lib/runapi/wan/resources/animate.rb
# frozen_string_literal: true

module RunApi
  module Wan
    module Resources
      # Transfers motion from a reference video onto a subject in the source image.
      # Use wan-2.2-animate-move to preserve the subject and animate its motion,
      # or wan-2.2-animate-replace to swap the subject with the reference video's subject.
      class Animate
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/wan/animate"
        RESPONSE_CLASS = Types::VideoTaskResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedVideoTaskResponse

        def initialize(http)
          @http = http
        end

        # Transfer motion and wait until complete.
        #
        # @param params [Hash] animation parameters
        # @return [RunApi::Wan::Types::CompletedVideoTaskResponse] completed animation result
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Create an animation task.
        #
        # @param params [Hash] animation parameters
        # @return [RunApi::Wan::Types::VideoTaskResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get animation status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::Wan::Types::VideoTaskResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["animate"], params)
        end
      end
    end
  end
end

File: python/src/runapi/wan/resources/animate.py
"""Wan animate resource."""

from __future__ import annotations

from typing import Any, Optional

from runapi.core import Resource, RequestOptions

from ..contract_gen import CONTRACT
from ..types import (
    CompletedVideoTaskResponse,
    VideoTaskResponse,
)


class Animate(Resource):
    """Animation operations (animate-move, animate-replace) with Wan models."""

    ENDPOINT = "/api/v1/wan/animate"

    RESPONSE_CLASS = VideoTaskResponse
    COMPLETED_RESPONSE_CLASS = CompletedVideoTaskResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Transfer motion and poll until it completes.

        Args:
            **params: animation parameters (model, ...).

        Returns:
            The completed (narrowed) animation response.
        """
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        """Create a animation task and return immediately with an id.

        Args:
            **params: animation parameters (model, ...).

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["animate"], compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        """Fetch the current status of a animation task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)
